rtl/src/main/scala/tensil/mem/DualPortMem.scala

```diff
@@ -149,11 +149,9 @@ class DualPortMem[T <: Data](
         }
       }
       case ChiselSyncReadMem => {
-        // Assumes that the memory space for each port is completely separate.
-        // This is currently true because we only use portA for data and portB
-        // for weights.
+        val mem = SyncReadMem(depth, gen, SyncReadMem.ReadFirst)
```

Comment:
I'm surprised this passed CI because as far as I understand, calling `mem.read()` twice causes undefined behavior.

From the docs:

```
  * Read-after-write behavior (when
  * a read and write to the same address are requested on the same cycle) is
  * undefined.
  *
  * @note when multiple conflicting writes are performed on a Mem element, the
  * result is undefined (unlike Vec, where the last assignment wins)
```

I tihnk what we actually need to do is implement a reg(vec()) based dual port mem. Wdyt?

Reply:
Do you think this might be ok since the same behavior is usually expected in hardware dual-port memory? (I've see people suggesting this implementation for Chisel when searching for solution.)

Is RegisterBank taking reg(vec()) approach? I tried it with XOR model and it crashed after long wait.

Reply:
Interesting, do you have a link to those suggestions? I'd like to dig into it a little because I struggled with this problem before and didn't find any good answers. The RegisterBank version is my attempt and as you found it doesnt seem to work, although I can't figure out why.

I suppose that since this one actually works then despite the warning in the docs we can just use it and merge this! Can you drop in a note explaining this context, so that if/when it stops working we know where to start looking?

Reply:
Here's the link to Chisel dual-port memory discussion https://github.com/chipsalliance/chisel3/issues/1788

(They actually raised the issue with impossibility for Yosis to infer Xilinx BRAM from Chisel SyncReadMem, but don't seem to have objections to the Chisel code itself.)